arrow-ipc/src/reader.rs

```diff
@@ -490,22 +490,44 @@ impl<'a> RecordBatchDecoder<'a> {
                     self.skip_field(field, &mut variadic_counts)?;
                 }
             }
-            assert!(variadic_counts.is_empty());
+
             arrays.sort_by_key(|t| t.0);
-            RecordBatch::try_new_with_options(
-                Arc::new(schema.project(projection)?),
-                arrays.into_iter().map(|t| t.1).collect(),
-                &options,
-            )
+
+            let schema = Arc::new(schema.project(projection)?);
+            let columns = arrays.into_iter().map(|t| t.1).collect::<Vec<_>>();
+
+            if self.skip_validation.get() {
+                unsafe {
```

Comment:
I think we typically add a comment to justify the use of any unsafe code. 
```suggestion
                // Safety: setting `skip_validation` requires `unsafe`, user assures data is valid
                unsafe {
```